### PR TITLE
[SPARK-13962][ML] spark.ml Evaluators should support other numeric types for label

### PR DESCRIPTION
**Pull request by BenFradet**
## What changes were proposed in this pull request?

Made BinaryClassificationEvaluator, MulticlassClassificationEvaluator and RegressionEvaluator accept all numeric types for label
## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by BenFradet** [comment]
pinging @MLnick

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala [hunk 1]
```diff
@@ -74,21 +74,14 @@ final class RegressionEvaluator @Since("1.4.0") (@Since("1.4.0") override val ui
...
+    SchemaUtils.checkNumericType(schema, $(labelCol))
 
     val predictionAndLabels = dataset
       .select(col($(predictionCol)).cast(DoubleType), col($(labelCol)).cast(DoubleType))
-      .rdd.
-      map { case Row(prediction: Double, label: Double) =>
-        (prediction, label)
+      .rdd
+      .map {
+        case Row(prediction: Double, label: Double) => (prediction, label)
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: minor but can this fit on one line? or have `.map { case Row ... }`

**Comment by MLnick** [comment]
LGTM. Will leave open a little while in case anyone else wants to take a look. @sethah?

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeTests.scala [hunk 2]
```diff
@@ -79,16 +79,20 @@ private[ml] object TreeTests extends SparkFunSuite {
    *              This must be non-empty.
    * @param numClasses  Number of classes label can take. If 0, mark as continuous.
    * @param labelColName  Name of the label column on which to set the metadata.
+   * @param featuresColName  Name of the features column
    * @return DataFrame with metadata
    */
-  def setMetadata(data: DataFrame, numClasses: Int, labelColName: String): DataFrame = {
+  def setMetadata(data: DataFrame,
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeTests.scala]: Follow Spark style here:

```scala
def setMetadata(
      data: DataFrame,
      numClasses: Int,
      labelColName: String,
      featuresColName: String): DataFrame = {
```

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala [hunk 3]
```diff
@@ -92,21 +111,28 @@ object MLTestingUtils extends SparkFunSuite {
     types
       .map(t => t -> df.select(col(labelColName).cast(t), col(featuresColName)))
       .map { case (t, d) =>
-        t -> TreeTests.setMetadata(d, 0, labelColName).withColumn(censorColName, lit(0.0))
+        t -> TreeTests.setMetadata(d, 0, labelColName, featuresColName)
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: minor: Not related to the changes in this PR, but I tend to prefer a single map statement like:

```scala
types.map { t =>
      val castDF = df.select(col(labelColName).cast(t), col(featuresColName))
      t -> TreeTests.setMetadata(castDF, 2, labelColName, featuresColName)
    }.toMap
```

This happens in the `genRegressionDFWithNumericLabelCol` as well.

**Comment by BenFradet** [comment]
@MLnick @sethah thanks for the reviews, will fix.

**Comment by sethah** [comment]
A couple minor syntax comments, other than that LGTM.

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala [hunk 4]
```diff
@@ -47,14 +48,32 @@ object MLTestingUtils extends SparkFunSuite {
...
+    val actuals = dfs.keys.filter(_ != DoubleType).map(t => evaluator.evaluate(dfs(t)))
+    actuals.foreach(actual => assert(expected === actual))
+
+    val dfWithStringLabels = sqlContext.createDataFrame(Seq(
+      ("0", 0d)
+    )).toDF("label", "prediction")
+    val thrown = intercept[IllegalArgumentException] {
+      evaluator.evaluate(dfWithStringLabels)
+    }
+    assert(thrown.getMessage contains
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: Minor style issue, but Spark code style prefers not to use infix notation (see https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide). Could you do `thrown.getMessage.contains(...)` instead? And also change the occurrence above in L57.

**Comment by MLnick** [comment]
@BenFradet just one more minor style comment, then I think this is ready to merge.

**Comment by BenFradet** [comment]
@MLnick will do

**Comment by MLnick** [comment]
@BenFradet Thanks! Merged to master. Thanks @sethah for the review.

**Pull request closed without merging**